# Исследование надежности заемщиков


## Описание проекта
Каждый банк производит оценку заемщиков для определения аккуратности/пунктуальности в погашении задолженности, эта оценка называется скоринг. <br>
Скоринг – используемая банками система оценки клиентов, в основе которой заложены статистические методы. Как правило, это компьютерная программа, куда вводятся данные потенциального заемщика. В ответ выдается результат – стоит ли предоставлять ему кредит. Название скоринг происходит от английского слова score, то есть «счет».

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Цель
Необходимо разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


## Описание данных
- `'children'` — количество детей в семье
- `'days_employed`' — общий трудовой стаж в днях
- `'dob_years`' — возраст клиента в годах
- `'education`' — уровень образования клиента
- `'education_id`' — идентификатор уровня образования
- `'family_status`' — семейное положение
- `'family_status_id`' — идентификатор семейного положения
- `'gender`' — пол клиента
- `'income_type`' — тип занятости
- `'debt`' — имел ли задолженность по возврату кредитов
- `'total_income`' — ежемесячный доход
- `'purpose`' — цель получения кредита

## План работы
* Откройте таблицу и изучите общую информацию о данных.
* Самостоятельно изучаем DataFrame.
* Предобработка данных.
* Исследуйте данные и ответьте на вопросы.
* Общий вывод.

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('E:\WORK\Образование\Аналитик данных\Dataset\Dataset\\Data_sprint4.csv')

In [2]:
data1 = data.copy(deep=True)

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Самостоятельно изучаем DataFrame

In [5]:
def data_analis(data=data, lenght=100, text=False):
    """
    Функция для первичного анализа DataFrame и формирования описательной части в Jupyter
    Передаваемые параметры: data - наш DataFrame, lenght - максимальное кол-во уникальных значений в столбце, text - надо ли формировать описательную часть
    
    """
    print('=====================================================================================================================')
    print('Анализ DataFrame')
    print('=====================================================================================================================')
    print('Этап 1. Первичный анализ.')
    print('=====================================================================================================================')
    display(data.sample(5, random_state =27))
    display(data.describe(include='all'))
    data.info()
    print('==================')
    print('Названия стобцов.')
    print('==================')
    print(data.columns)
    print('==================')
    print('Явные пропуски.')
    print('==================')
    print(data.isna().mean())
    print('==================')
    print('Явные дубликаты.')
    print('==================')
    print(data.duplicated().sum())
    print('==================')
    for i in data.columns:
        if len(data[i].unique()) <= lenght:
            print('======================================================')
            print('Уникальные значения и их кол-во в столбце:', i)
            print('======================================================')
            print(data.sort_values(by=i).groupby(i)[i].count())
    print('======================================================')
    print('Основные параметры числовых значений')
    print('======================================================')
    display(data.describe())
    if text:
        print('\n\n\n')
        print('Таблица для оформления анализа даннных DF.')
        print('### Анализ качества полученных данных.')
        print('В DataFrame:\n<br>- столбцов: ', len(data.columns), '\n<br>- строк: ', len(data), '\n<br>- явных дубликатов: ', data.duplicated().sum(), '\n<br>- доля пропусков в столбцах:\n<br>', data.isna().mean()[data.isna().mean() != 0],'\n<br>', sep='',end='')
        print('- название столбцов:\n', sep='',end='')
        for i in data.columns:
            print('<br>|', i,'|', sep='', end='')
        print('\n\nПроанализируем качество данных в столбцах:')
        for i in data.columns:
            x="'"
            print('- `', x, i, x, '`: Название столбца: правильное/не правильное. Значение: . Проблемы: . Наблюденние: . Рекомендации: .', sep='')            

In [6]:
data_analis(data, text=True)

Анализ DataFrame
Этап 1. Первичный анализ.


children  days_employed  dob_years education  education_id  \
1018          1  356170.475243         64   среднее             1   
11876         0  396541.413988         72    высшее             0   
2783          1   -1702.357343         49   среднее             1   
8913          2   -1948.471496         37   среднее             1   
5781          0    -628.568068         23    высшее             0   

               family_status  family_status_id gender  income_type  debt  \
1018          вдовец / вдова                 2      F    пенсионер     0   
11876        женат / замужем                 0      M    пенсионер     0   
2783         женат / замужем                 0      M    сотрудник     0   
8913        гражданский брак                 1      M  госслужащий     0   
5781   Не женат / не замужем                 4      F    компаньон     0   

        total_income                           purpose  
1018   101473.954477           покупка жилья для семьи  
11876  159358.701867                        автомобиль  
2783   308147.977006  строительство жилой недвижимости  
8913   191526.819211  сделка с подержанным автомобилем  
5781   123168.447444             заняться образованием

children  days_employed     dob_years education  education_id  \
count   21525.000000   19351.000000  21525.000000     21525  21525.000000   
unique           NaN            NaN           NaN        15           NaN   
top              NaN            NaN           NaN   среднее           NaN   
freq             NaN            NaN           NaN     13750           NaN   
mean        0.538908   63046.497661     43.293380       NaN      0.817236   
std         1.381587  140827.311974     12.574584       NaN      0.548138   
min        -1.000000  -18388.949901      0.000000       NaN      0.000000   
25%         0.000000   -2747.423625     33.000000       NaN      1.000000   
50%         0.000000   -1203.369529     42.000000       NaN      1.000000   
75%         1.000000    -291.095954     53.000000       NaN      1.000000   
max        20.000000  401755.400475     75.000000       NaN      4.000000   

          family_status  family_status_id gender income_type          debt  \
count             21525      21525.000000  21525       21525  21525.000000   
unique                5               NaN      3           8           NaN   
top     женат / замужем               NaN      F   сотрудник           NaN   
freq              12380               NaN  14236       11119           NaN   
mean                NaN          0.972544    NaN         NaN      0.080883   
std                 NaN          1.420324    NaN         NaN      0.272661   
min                 NaN          0.000000    NaN         NaN      0.000000   
25%                 NaN          0.000000    NaN         NaN      0.000000   
50%                 NaN          0.000000    NaN         NaN      0.000000   
75%                 NaN          1.000000    NaN         NaN      0.000000   
max                 NaN          4.000000    NaN         NaN      1.000000   

        total_income  purpose  
count   1.935100e+04    21525  
unique           NaN       38  
top              NaN  свадьба  
freq             NaN      797  
mean    1.674223e+05      NaN  
std     1.029716e+05      NaN  
min     2.066726e+04      NaN  
25%     1.030532e+05      NaN  
50%     1.450179e+05      NaN  
75%     2.034351e+05      NaN  
max     2.265604e+06      NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
Названия стобцов.
Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt'

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06





Таблица для оформления анализа даннных DF.
### Анализ качества полученных данных.
В DataFrame:
<br>- столбцов: 12
<br>- строк: 21525
<br>- явных дубликатов: 54
<br>- доля пропусков в столбцах:
<br>days_employed    0.100999
total_income     0.100999
dtype: float64
<br>- название столбцов:
<br>|children|<br>|days_employed|<br>|dob_years|<br>|education|<br>|education_id|<br>|family_status|<br>|family_status_id|<br>|gender|<br>|income_type|<br>|debt|<br>|total_income|<br>|purpose|

Проанализируем качество данных в столбцах:
- `'children'`: Название столбца: правильное/не правильное. Значение: . Проблемы: . Наблюденние: . Рекомендации: .
- `'days_employed'`: Название столбца: правильное/не правильное. Значение: . Проблемы: . Наблюденние: . Рекомендации: .
- `'dob_years'`: Название столбца: правильное/не правильное. Значение: . Проблемы: . Наблюденние: . Рекомендации: .
- `'education'`: Название столбца: правильное/не правильное. Значение: . Проблемы: . Наблюденние: . Рекомендации: .
- `

### Анализ качества полученных данных.
В DataFrame:
<br>- столбцов: 12
<br>- строк: 21525
<br>- явных дубликатов: 54
<br>- доля пропущеных значенний в столбцах:
<br>days_employed    0.100999
total_income     0.100999
dtype: float64
<br>- Название столбцов:
<br>|children|<br>|days_employed|<br>|dob_years|<br>|education|<br>|education_id|<br>|family_status|<br>|family_status_id|<br>|gender|<br>|income_type|<br>|debt|<br>|total_income|<br>|purpose|

Проанализируем качество данных в столбцах:
- `'children'`: Название столбца: правильное. Значение: количество детей в семье. Проблемы: есть значенние -1 (47 шт.) и значение 20 (76 шт.), вероятно это 1 и 2. Наблюдение: У 75% не более 1 ребенка. Рекомендации: 1) т.к. коолчиество ошибок не велико, можно строки дроопнуть; 2) дать рекомендацию разработчику формы, запретить ввод отрицательных величин и дать алерт при введении значений больше 4, чтобы специалист обратил на это внимание.
- `'days_employed'`: Название столбца: правильное. Значение: общий трудовой стаж в днях. Проблемы: 1) есть отрицательные значения есть очень большие значения, переводятся как дни работы, но если перевести в года, то столько не живут; 2) есть пропуски. Рекомендации: проверить неправдоподобные значения и перевести в положительные, разобраться с пропусками.
- `'dob_years'`: Название столбца: правильное. Значение: возраст клиента в годах. Проблемы: есть нулевые значения (101 шт.). Наблюденние: максимальноое кол-во заемщиков в возрасте 26-60 лет. Рекомендации: 1) 0-ые значения можно не удалять т.к. роль этих данных не существенна; 2) дать рекомендацию разработчику формы, ввести порог на минимальный возраст, выдавать алерт, если возраст больше 70.
- `'education'`: Название столбца: правильное. Значение: уровень образования клиента. Проблемы: 15 уникальных значенний, значения дублируются, т.к. находятся в разном регистре. Наблюденние: больше всего заемщиков с средним обрразованиеем, меньше всего с ученой степенью. Рекомендации: 1) привести к одному регистру; 2) дать рекомендацию разработчику формы, данные по образование через выпадающий список.
- `'education_id'`: Название столбца: правильное. Значение:  идентификатор уровня образования (например: 0 - высшее, 1 - среднее). Проблемы: нет. Наблюденние: больше всего заемщиков с средним образованием, меньше всего с учееной степенью. Рекомендации: нет.
- `'family_status'`: Название столбца: правильное. Значение: семейное положение. Проблемы: разный регистр. Наблюденние: . Рекомендации: 1) привести к одному регистру; 2) дать рекомендацию разработчику формы, данные по семейнному положению через выпадающий список.
- `'family_status_id'`: Название столбца: правильное. Значение: идентификатор семейного положения (например: 0 - женат / замужем, 1 - гражданский брак). Проблемы: нет. Наблюденние: 1) значения могут кочевать из одного типа в другой или могут быть условнно одинаковыми (например вдовец, не женат, гражданский брак), но они относятся к разным группам что может влиять на скоринг; 2) больше всегго заемщиков среди женатых, меньше среди вдовцов. Рекомендации: уточнить, как проверяется семейный статус, насколько ему можно верить.
- `'gender'`: Название столбца: правильное. Значение: пол клиента. Проблемы: 1) верхний регистр; 2) есть одно знначение с нееизвестным полом. Наблюденние: среди заемщиков в 2 раза больше женщин. Рекомендации: 1) привести к нижнему регистру, данные с непонятнным полом можно оставить т.к. данная информация не критичнна; 2) дать рекомендацию разработчику формы, данные по полу получать через выпадающий список.
- `'income_type'`: Название столбца: правильное. Значение: тип занятости. Проблемы: не понятно как проверяются данные, не мигрируют ли типы между собой, практически нулевое кол-во значний у безработный/в декрете/предприниматель/студент, не понятно кто такие компаньены и как вообще проверяются данные, насколько им можно верить. Наблюденние: больше всего заемщиков в категории сотрудник, меньше всего в госслужащий, вероятно из-за разницы в их общем кол-ве. Рекомендации: 1) убрать не влияющие на статистику данные; 2) дать рекомендацию разработчику формы, данные по типу дохода получать через выпадающий список.
- `'debt'`: Название столбца: правильное. Значение: имел ли задолженность по возврату кредитов. Проблемы: нет. Наблюденние: значений с статусом 0 болше чем 1 в 10 раз. Рекомендации: нет.
- `'total_income'`: Название столбца: правильное. Значение: ежемесячный доход. Проблемы: данные вещественные и есть пропуски. Наблюденние: 75% людей зарабатываю до 200 т.р. в месяц. Рекомендации: привести данные к целым, разобраться с пропусками.
- `'purpose'`: Название столбца: правильное. Значение: цель получения кредита. Проблемы: 38 уникальных значениий, значения в одном регистре, но нет жесткой формы, необходимо группировать. Наблюденние: необходимо группировать. Рекомендации: 1) группировать данные; 2) дать рекомендацию разработчику формы, данные по цели кредита получать через выпадающий список.

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [8]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t) & (data['total_income'].isna() == False), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [9]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

Уберем одну запись у которой gender XNA.

In [11]:
data = data.loc[data['gender'] != 'XNA']

Уберем 100 записей с возрастом 0.

In [12]:
data = data.loc[data['dob_years'] != 0]

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [13]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [14]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [15]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [16]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna() == False), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [17]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [18]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [19]:
for i in data.columns:
    if data[i].dtypes == 'object':
        data[i] = data[i].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [20]:
data.duplicated().sum()

71

In [21]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [22]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [23]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [24]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [25]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [26]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [27]:
print('Всего заемщиков:', len(data),'. Заемщиков у которых менее 3-ех детей:{:.2%}'.format
      (len(data[data['children'] < 3])/len(data)), sep='')
display(data[data['children'] < 3].pivot_table(index=['children'], values='debt', aggfunc=['mean','count']))

Всего заемщиков:21230. Заемщиков у которых менее 3-ех детей:98.22%


mean  count
              debt   debt
children                 
0         0.075458  14021
1         0.092028   4792
2         0.095145   2039

**Вывод:** Т.к. заемщики у которых более 3-ех детей составляют всего лишь ~2% от всех заемщиков, то считаем такую выборку не репрезентативной, соответственно ей можно пренебречь. Самая многочисленная группа заемщиков не имеет детей, у этой группы самый маленький процент невозврата, разница с остальными группами, среди тех кто не оплатил вовремя из этой группы, составляет до 20% и до 2% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между количеством детей и вероятностью возврата кредита в срок, а именно: **люди с меньшим кол-ом детей чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, есть подозрение, что если бы заемщиков с детьми было бы больше, то результаты были бы схожи.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [28]:
display(data.pivot_table(index='family_status', values='debt', aggfunc=['mean','count']).sort_values(by=( 'mean', 'debt')))

mean  count
                           debt   debt
family_status                         
вдовец / вдова         0.065539    946
в разводе              0.071247   1179
женат / замужем        0.075575  12213
гражданский брак       0.093142   4112
не женат / не замужем  0.097842   2780

**Вывод:** Самая многочисленная группа заемщиков находится в браке, у этой группы средний процент невозврата. Самый низкий процент вдов(цов) разница с остальными группами, среди тех кто не оплатил вовремя из этой группы, составляет до 33% и до 3% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между семейным положением и вероятностью возврата кредита в срок, а именно: **люди которые были или находятся в браке чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
display(data.pivot_table(index='total_income_category', values='debt', aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
                           debt   debt
total_income_category                 
D                      0.060519    347
B                      0.070784   4987
A                      0.080000     25
C                      0.084926  15849
E                      0.090909     22

In [30]:
display(data[(data['total_income_category'] != 'E') & (data['total_income_category'] != 'A') & (data['total_income_category'] != 'D')].pivot_table(index='total_income_category', values='debt', aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
                           debt   debt
total_income_category                 
B                      0.070784   4987
C                      0.084926  15849

**Вывод:** Из-за количества, группами E,A,D можно пренебречь. Самый низкий процент не возврата в срок у людей из группы B, разница с остальными группами, среди тех кто не оплатил вовремя из этой группы, составляет до 17% и до 1% из общего количества в группах.
Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между ежемесячным доходом и вероятностью возврата кредита в срок, а именно: **люди с большим достатком чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [31]:
display(data.pivot_table(index='purpose_category', values='debt', aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
                              debt   debt
purpose_category                         
операции с недвижимостью  0.072596  10703
проведение свадьбы        0.078730   2299
получение образования     0.092947   3970
операции с автомобилем    0.093236   4258

**Вывод:** Самая многочисленная группа заемщиков берет кредит на недвижимость, у этой группы самый низкий процент не возврата кредита в срок, разница с остальными группами, среди тех кто не оплатил вовремя из этой группы, составляет до 22% и до 3% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между целью кредита и вероятностью возврата кредита в срок, а именно: **люди, которые берут кредит на более глобальные цели (жилье, семья) чаще возвращают кредит в срок.**  Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

### Проверка других зависимостей

#### 3.5 Как образование влияет на возврат кредита в срок?

In [32]:
display(data.pivot_table(index='education', values='debt', aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
                         debt   debt
education                           
ученая степень       0.000000      6
высшее               0.053351   5192
среднее              0.089795  15012
неоконченное высшее  0.092141    738
начальное            0.109929    282

**Вывод:** Из-за количества, группой с ученой степенью и начальным образованием можно пренебречь. Самая многочисленная группа заемщиков имеет среднее образование, у этой группы средний процент невозврата кредита в срок. Самый низкий процент невозврата в срок у людей с высшим образованием, разница с остальными группами, среди тех, кто не оплатил вовремя из этой группы, составляет до 42% и до 4% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между образованием заемщика и вероятностью возврата кредита в срок, а именно: **более образованные люди чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

#### 3.6 Как возраст влияет на возврат кредита в срок?

In [33]:
display(data.pivot_table(index='dob_years' ,values='debt',\
        aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean count
               debt  debt
dob_years                
75         0.000000     1
74         0.000000     6
73         0.000000     8
71         0.017857    56
61         0.028409   352
66         0.043956   182
63         0.044776   268
70         0.046154    65
64         0.046512   258
69         0.048193    83
62         0.051873   347
55         0.052154   441
58         0.052863   454
67         0.053892   167
51         0.054054   444
52         0.057971   483
60         0.058981   373
72         0.060606    33
49         0.061386   505
48         0.061798   534
54         0.065539   473
57         0.066225   453
56         0.066946   478
59         0.068182   440
42         0.071066   591
19         0.071429    14
65         0.072539   193
44         0.075786   541
53         0.076586   457
35         0.076672   613
39         0.077058   571
47         0.077568   477
36         0.077899   552
40         0.078203   601
20         0.078431    51
45         0.079108   493
23         0.080000   250
68         0.080808    99
41         0.081531   601
46         0.081545   466
43         0.086444   509
50         0.090551   508
24         0.091603   262
33         0.095321   577
37         0.096226   530
27         0.100000   490
32         0.100990   505
30         0.101313   533
34         0.105882   595
38         0.107744   594
28         0.113772   501
26         0.115764   406
29         0.119926   542
25         0.120787   356
21         0.127273   110
22         0.136612   183
31         0.136937   555

Зависимость есть но таблица плохочитаемая, необходимо категоризировать данные. Разобьем всех клиентов на поколения.

In [34]:
for i in data['dob_years'].unique():    
    if 1944 > 2022 - i >= 1925:
        data.loc[data['dob_years'] == i, 'generation'] = 'молчаливое'
    if 1967 > 2022 - i >= 1944:
        data.loc[data['dob_years'] == i, 'generation'] = 'бэби бумеры'
    if 1984 > 2022 - i >= 1967:
        data.loc[data['dob_years'] == i, 'generation'] = 'X'
    if 2000 > 2022 - i >= 1984:
        data.loc[data['dob_years'] == i, 'generation'] = 'Y'
    if 2015 > 2022 - i >= 2000:
        data.loc[data['dob_years'] == i, 'generation'] = 'Z'
    
display(data.pivot_table(index='generation' ,values='debt',\
        aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean count
                 debt  debt
generation                 
бэби бумеры  0.055375  4316
X            0.072685  8695
Y            0.102913  7861
Z            0.122905   358

**Вывод:** Из-за количества, группой с людьми поколения Z можно пренебречь. Самая многочисленная группа заемщиков относится к поколению X, у этой группы средний процент невозврата кредита в срок. Самый низкий процент не возврата в срок у поколения беби бумеров, разница с остальными группами, среди тех кто не оплатил вовремя из этой группы, составляет до 46% и до 5% из общего количества в группах. <br> Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между возрастом и вероятностью возврата кредита в срок, а именно: **люди в возрасте существенно чаще возвращают кредит в срок.**

#### 3.7 Как тип занятости влияет на возврат кредита в срок?

In [35]:
display(data.pivot_table(index='income_type' ,values='debt',\
        aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
                     debt   debt
income_type                     
предприниматель  0.000000      2
студент          0.000000      1
пенсионер        0.056698   3792
госслужащий      0.059516   1445
компаньон        0.074214   5026
сотрудник        0.095612  10961
безработный      0.500000      2
в декрете        1.000000      1

**Вывод:** Из-за количества, группой предпринимателей, студентов, безработных и находящихся в декрете можно пренебречь. Самая многочисленная группа заемщиков - сотрудники, у этой группы самый высокий процент невозврата кредита в срок. Самый низкий процент невозврата в срок у пенсионеров, разница с остальными группами, среди тех, кто не оплатил вовремя из этой группы, составляет до 41% и до 4% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между типом занятости и вероятностью возврата кредита в срок, а именно: **люди имеющие постоянный, гарантированный доход от государства чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

#### 3.9 Как пол влияет на возврат кредита в срок?

In [36]:
display(data.pivot_table(index='gender' ,values='debt',\
        aggfunc=['mean','count']).sort_values(by=('mean', 'debt')))

mean  count
            debt   debt
gender                 
f       0.070180  14021
m       0.102649   7209

**Вывод:** Самая многочисленная группа заемщиков - женщины (больше в 2 раза). Самый низкий процент невозврата в срок у женщин, разница с остальными группами, среди тех, кто не оплатил вовремя из этой группы, составляет до 32% и до 3% из общего количества в группах. <br>Таким образом, исходя из имеющихся данных, можно сказать, что прослеживается зависимость между полом заемщика и вероятностью возврата кредита в срок, а именно: **женщины чаще возвращают кредит в срок.** Т.к. группы существенно отличаются по количеству, основное влияние на общее кол-во невозвратов в срок оказывает самая многочисленная группа.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

In [37]:
print('Количество пропусков в столбце days_employed:', data1['days_employed'].isna().sum(),'\nв столбце total_income:',data1['days_employed'].isna().sum(), 
      '\nв двух столбцах одновременно:', len(data1[(data1['days_employed'].isna()) & (data1['total_income'].isna())]),
      '\nв двух столбцах одновременно и при значении в debt=0:', len(data1[(data1['days_employed'].isna()) & (data1['total_income'].isna()) & (data1['debt'] == 0)]))
display(data1[(data1['days_employed'].isna()) & (data1['total_income'].isna() & (data1['debt'] == 1))].sample(10, random_state=27))

Количество пропусков в столбце days_employed: 2174 
в столбце total_income: 2174 
в двух столбцах одновременно: 2174 
в двух столбцах одновременно и при значении в debt=0: 2004


children  days_employed  dob_years education  education_id  \
20314         0            NaN         27   среднее             1   
8301          0            NaN         36    высшее             0   
19251         0            NaN         38   среднее             1   
16394         2            NaN         36   среднее             1   
9153          0            NaN         57   среднее             1   
15592         1            NaN         38   среднее             1   
18556         1            NaN         25   среднее             1   
20354         0            NaN         59   среднее             1   
10563         0            NaN         72   Среднее             1   
3888          2            NaN         34   СРЕДНЕЕ             1   

               family_status  family_status_id gender income_type  debt  \
20314        женат / замужем                 0      M   сотрудник     1   
8301        гражданский брак                 1      M   компаньон     1   
19251        женат / замужем                 0      M   сотрудник     1   
16394         вдовец / вдова                 2      F   пенсионер     1   
9153         женат / замужем                 0      M   пенсионер     1   
15592        женат / замужем                 0      F   сотрудник     1   
18556       гражданский брак                 1      F   компаньон     1   
20354        женат / замужем                 0      F   пенсионер     1   
10563  Не женат / не замужем                 4      F   пенсионер     1   
3888        гражданский брак                 1      F   сотрудник     1   

       total_income                           purpose  
20314           NaN  строительство жилой недвижимости  
8301            NaN                           свадьба  
19251           NaN           покупка жилья для сдачи  
16394           NaN                   свой автомобиль  
9153            NaN   операции со своей недвижимостью  
15592           NaN           приобретение автомобиля  
18556           NaN           покупка жилья для семьи  
20354           NaN                     покупка жилья  
10563           NaN           приобретение автомобиля  
3888            NaN             на проведение свадьбы

In [38]:
for i in data1.columns:
    print(i, ':', data1[(data1['days_employed'].isna()) & (data1['total_income'].isna())][i].unique())

children : [ 0  1  2 -1  3  4 20  5]
days_employed : [nan]
dob_years : [65 41 63 50 54 21 52 32 35 34 44 47 46 29 27 39 62 55 26 30 31 23 58 60
 40 66 33 56 69 59 57 45 53 48 51 37 43 24 28 67 64 25 49 42 61 36 70 38
 20  0 68 19 22 71 72 73]
education : ['среднее' 'высшее' 'СРЕДНЕЕ' 'неоконченное высшее' 'Среднее'
 'Неоконченное высшее' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'начальное' 'ВЫСШЕЕ'
 'Начальное' 'НАЧАЛЬНОЕ']
education_id : [1 0 2 3]
family_status : ['гражданский брак' 'женат / замужем' 'Не женат / не замужем'
 'вдовец / вдова' 'в разводе']
family_status_id : [1 0 4 2 3]
gender : ['M' 'F']
income_type : ['пенсионер' 'госслужащий' 'компаньон' 'сотрудник' 'предприниматель']
debt : [0 1]
total_income : [nan]
purpose : ['сыграть свадьбу' 'образование' 'строительство жилой недвижимости'
 'сделка с подержанным автомобилем'
 'операции с коммерческой недвижимостью' 'покупка жилья для семьи' 'жилье'
 'операции с жильем' 'на проведение свадьбы' 'покупка жилой недвижимости'
 'профильное обра

*Ответ:* Пропуски в столбцах `'days_employed'` и `'total_income'` связаны между собой, т.к. они находятся в одних и тех же строках, большинство из них имеют в столбце `'debt'` 0, что не является аномалией, число не возвратов по разным группам до 10%.<br> Т.к. явной закономерности с другими столбцами нет, вероятно пропуск допущен в одном из столбцов, а во втором столбце система сама сделала пропуск. Скорее всего, некоторые клиенты отказывались называть свой заработок.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Для получения значений близких к реальным, необходимо отсечь пики, сильно искажающие средние значения и учесть данные с которыми есть корреляция. Поэтому мы брали медианные значения из группы значений соответствующих типу занятости клиента.

### Шаг 4: общий вывод.

В данном исследовании мы работали со следующими данными:
- `'children'`: Значение: количество детей в семье. В данных обнаружены проблемы: есть значение -1 (47 шт.) и значение 20 (76 шт.), вероятно это 1 и 2.
- `'days_employed'`: Значение: общий трудовой стаж в днях. В данных обнаружены проблемы: 1) есть отрицательные значения есть очень большие значения, переводятся как дни работы, но если перевести в года, то столько не живут; 2) есть пропуски. 
- `'dob_years'`: Значение: возраст клиента в годах. В данных обнаружены проблемы: есть нулевые значения (101 шт.). 
- `'education'`: Значение: уровень образования клиента. В данных обнаружены проблемы: 15 уникальных значении, значения дублируются, т.к. находятся в разном регистре.
- `'education_id'`: Значение: идентификатор уровня образования (например: 0 - высшее, 1 - среднее). 
- `'family_status'`: Значение: семейное положение. В данных обнаружены проблемы: разный регистр. 
- `'family_status_id'`: Значение: идентификатор семейного положения. В данных обнаружены проблемы: нет. Наблюдение: 1) значения могут кочевать из одного типа в другой или могут быть условно одинаковыми (например, вдовец, не женат, гражданский брак), но они относятся к разным группам что может влиять на скоринг; 2) больше всего заемщиков среди женатых, меньше среди вдовцов.
- `'gender'`: Значение: пол клиента. В данных обнаружены проблемы: 1) верхний регистр; 2) есть одно значение с неизвестным полом. Наблюдение: среди заемщиков в 2 раза больше женщин. 
- `'income_type'`: Значение: тип занятости. В данных обнаружены проблемы: не понятно, как проверяются данные, не мигрируют ли типы между собой, практически нулевое кол-во знаний у безработный/в декрете/предприниматель/студент, не понятно кто такие компаньоны и как вообще проверяются данные, насколько им можно верить. 
- `'debt'`: Значение: имел ли задолженность по возврату кредитов. 
- `'total_income'`: Значение: ежемесячный доход. В данных обнаружены проблемы: данные вещественные и есть пропуски. 
- `'purpose'`: Значение: цель получения кредита. В данных обнаружены проблемы: 38 уникальных значении, значения в одном регистре, но нет жесткой формы, необходимо группировать.

В процессе предобработки, различными методами все обнаруженные в данных проблемы были устранены.

**Итоги:** <br>
В данном исследовании мы убедились, что существует зависимость невозврата клиентом кредита в срок от:
- `количества детей` - **люди с меньшим кол-ом детей чаще возвращают кредит в срок.**
- `семейного положения` - **люди которые были или находятся в браке чаще возвращают кредит в срок.**
- `ежемесячный доход` - **люди с большим достатком чаще возвращают кредит в срок.**
- `цели кредита` - **люди, которые берут кредит на более глобальные цели (жилье, семья) чаще возвращают кредит в срок.**
- `образования` - **более образованные люди чаще возвращают кредит в срок.**
- `поколения` - **люди в возрасте существенно чаще возвращают кредит в срок.**
- `типа занятости` - **люди имеющие постоянный, гарантированный доход от государства чаще возвращают кредит в срок.**
- `пола` - **женщины чаще возвращают кредит в срок.**
<br><br>

Для скоринга необходимо придать параметрам веса и добавить в общую формулу. Если бы были данные по невозвратным кредитам, можно было бы уточнить веса, прогнав формулу по существующим данным, отслеживая кол-во невозвратных кредитов. Т.к. просрочка не является убытком для банка.

### Рекомендации по сбору данных

- `'children'`: Рекомендации: запретить ввод отрицательных величин и дать алерт при введении значений больше 4, чтобы специалист обратил на это внимание.
- `'days_employed'`: Рекомендации: проверить не правдоподобные значения и перевести в положительные, разобраться с пропусками. Как считается стаж?
- `'dob_years'`: Рекомендации: ввести порог на минимальный возраст, выдавать алерт, если возраст больше 70.
- `'education'`: Рекомендации: данные по образованию получать через выпадающий список.
- `'family_status'`: Рекомендации: данные по семейному положению через выпадающий список.
- `'family_status_id'`: Рекомендации: уточнить, как проверяется семейный статус, насколько ему можно верить.
- `'gender'`: Рекомендации: данные по полу получать через выпадающий список.
- `'income_type'`: Рекомендации: данные по типу дохода получать через выпадающий список.
- `'total_income'`: Рекомендации: уточнить причину появления пропусков.
- `'purpose'`: Рекомендации: данные по цели кредита получать через выпадающий список.